browser/net/brave_ad_block_tp_network_delegate_helper.cc

```diff
@@ -75,17 +151,24 @@ void OnBeforeURLRequestAdBlockTP(
   }
   DCHECK_NE(ctx->request_identifier, 0UL);
 
-  g_brave_browser_process->ad_block_service()->GetTaskRunner()
-      ->PostTaskAndReply(FROM_HERE,
-                         base::BindOnce(&ShouldBlockAdOnTaskRunner, ctx),
-                         base::BindOnce(&OnShouldBlockAdResult, next_callback,
-                                        ctx));
+  auto* rph = content::RenderProcessHost::FromID(ctx->render_process_id);
+  if (!rph) {
+    // This case occurs for the top-level request. It's okay to ignore the
+    // canonical name here since the top-level request can't be blocked
+    // anyways.
+    g_brave_browser_process->ad_block_service()
+        ->GetTaskRunner()
+        ->PostTaskAndReply(
+            FROM_HERE, base::BindOnce(&ShouldBlockAdOnTaskRunner,
+                ctx, base::Optional<std::string>()),
+            base::BindOnce(&OnShouldBlockAdResult, next_callback, ctx));
+  } else {
+     new AdblockCnameResolveHostClient(next_callback, ctx);
```

Comment:
rather than passing the `next_callback`, I would pass https://github.com/brave/brave-core/pull/6640/files#diff-5a17ec06c6e281e744cfa03e9d7561a6R164 so you don't have to duplicate it here https://github.com/brave/brave-core/pull/6640/files#diff-5a17ec06c6e281e744cfa03e9d7561a6R77

Reply:
I would also partially bind https://github.com/brave/brave-core/pull/6640/files#diff-5a17ec06c6e281e744cfa03e9d7561a6R162 `base::BindOnce(&ShouldBlockAdOnTaskRunner,
                ctx)` and pass that as well so `AdblockCnameResolveHostClient` doesn't have to stay in sync with any changes here

Reply:
We ended up keeping the logic only in `OnGetCnameResult`, renaming that to `ShouldBlockAdWithOptionalCname`, and calling it directly with the empty `Optional` if there's no render process host.